rust/kernel/platform_driver.rs

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use crate::error::{Error, KernelResult};
+use crate::{bindings, c_types, CStr};
+use alloc::boxed::Box;
+use core::marker::PhantomPinned;
+use core::mem::transmute;
+use core::pin::Pin;
+
+extern "C" {
+    #[allow(improper_ctypes)]
+    fn rust_helper_platform_get_drvdata(
+        pdev: *const bindings::platform_device,
+    ) -> *mut c_types::c_void;
+
+    #[allow(improper_ctypes)]
+    fn rust_helper_platform_set_drvdata(
+        pdev: *mut bindings::platform_device,
+        data: *mut c_types::c_void,
+    );
+}
+
+unsafe extern "C" fn probe_callback<T: PlatformDriver>(
+    pdev: *mut bindings::platform_device,
+) -> c_types::c_int {
+    let f = || {
+        let drv_data = T::probe(&mut PlatformDevice::new(pdev))?;
+        let drv_data = Box::try_new(drv_data)?;
```

Comment:
Instead of forcing the use of `Box`, we should use `PointerWrapper` (similarly to `FileOperations::open`) for two reasons:
1. Implementers know where the final location is of their data is, so they can for example initialise self-referential structures like mutexes. The way the code is now, this move will mess up self-references.
2. Implementers can choose which wrapper to use (`Arc`, `Box`, `Ref`, pinned versions of these, or they can implement their own).

Reply:
Very good point ! I got stuck realizing that returning `KernelResult<Self::DrvData>` would not allow self-referential structures (see the "pin this?" comment in the code below). But I think your suggestion is an elegant solution.

Reply:
I noticed that I kind-of duplicated existing `file_operations` infrastructure to make the `PlatformDevice`/`Device` abstractions - I even gave it the same name, `PointerWrapper` !! I'm assuming it makes sense to embrace & extend Wedson's `PointerWrapper`, and make it generally available so it can be re-used here?

Reply:
Actually, see #144 and #185 ;D